kong/router/fields.lua

```diff
@@ -0,0 +1,322 @@
+    end,
+
+    ["http.path"] =
+    function(v, params, c)
+      return c:add_value("http.path", params.uri)
+    end,
+
+    ["http.host"] =
+    function(v, params, c)
+      return c:add_value("http.host", params.host)
+    end,
+
+    ["tls.sni"] =
+    function(v, params, c)
+      return c:add_value("tls.sni", params.sni)
+    end,
+
+    ["net.protocol"] =
+    function(v, params, c)
+      return c:add_value("net.protocol", params.scheme)
+    end,
+
+    ["net.port"] =
+    function(v, params, c)
+      return c:add_value("net.port", params.port)
+    end,
+
+    ["http.headers."] =
+    function(v, params, c)
+      local headers = params.headers
+      if not headers then
+        return true
+      end
+
+      for _, h in ipairs(v) do
+        local v = headers[h]
+        local f = "http.headers." .. h
+
+        if type(v) == "string" then
+          local res, err = c:add_value(f, v)
+          if not res then
+            return nil, err
+          end
+
+        elseif type(v) == "table" then
+          for _, v in ipairs(v) do
+            local res, err = c:add_value(f, v)
+            if not res then
+              return nil, err
+            end
+          end
+        end -- if type(v)
+      end   -- for ipairs(v)
+
+      return true
+    end,
+
+    ["http.queries."] =
+    function(v, params, c)
+      local queries = params.queries
+      if not queries then
+        return true
+      end
+
+      for _, n in ipairs(v) do
+        local v = queries[n]
+        local f = "http.queries." .. n
+
+        -- the query parameter has only one value, like /?foo=bar
+        if type(v) == "string" then
+          local res, err = c:add_value(f, v)
+          if not res then
+            return nil, err
+          end
+
+        -- the query parameter has no value, like /?foo,
+        -- get_uri_arg will get a boolean `true`
+        -- we think it is equivalent to /?foo=
+        elseif type(v) == "boolean" then
+          local res, err = c:add_value(f, "")
+          if not res then
+            return nil, err
+          end
+
+        -- multiple values for a single query parameter, like /?foo=bar&foo=baz
+        elseif type(v) == "table" then
+          for _, v in ipairs(v) do
+            local res, err = c:add_value(f, v)
+            if not res then
+              return nil, err
+            end
+          end
+        end -- if type(v)
+      end   -- for ipairs(v)
+
+      return true
+    end,
+}
+
+
+function get_cache_key(fields, params)
+  for field, value in pairs(fields) do
+    local func = CACHE_KEY_FUNCS[field]
+
+    if func then
+      func(value, params, str_buf)
+    end
+  end
+
+  if not fields["http.host"] then -- preserve_host
+    CACHE_KEY_FUNCS["http.host"](true, params, str_buf)
+  end
+
+  if not fields["http.path"] then -- 05-proxy/02-router_spec.lua:1329
+    CACHE_KEY_FUNCS["http.path"](true, params, str_buf)
+  end
+
+  return str_buf:get()
+end
```

Comment:
Depending on an undefined behavior is not a good idea (, especially if it doesn't bring a huge benefit.

```lua
local tbl = {
    a = 0,
    b = 1,
    c = 2,
    d = 3,
    e = 4,
    f = 5,
    g = 6,
    h = 7,
    i = 8,
}

for k, v in pairs(tbl) do
    print(string.format("%s: %s", k, v))
end

print("==========================")

local keys = {}
for _ = 1, 1000 do
    table.insert(keys, tostring(math.random()))
end

for _, k in ipairs(keys) do
    tbl[k] = math.random()  -- rehashing might happens here
end

for _, k in ipairs(keys) do
    tbl[k] = nil -- rehashing might happens here
end

collectgarbage("collect") -- rehashing happens here

for k, v in pairs(tbl) do
    print(string.format("%s: %s", k, v))
end
```

```bash
$ resty test.lua
h: 7
d: 3
a: 0
e: 4
i: 8
f: 5
b: 1
c: 2
g: 6
==========================
f: 5
b: 1
c: 2
g: 6
h: 7
e: 4
a: 0
d: 3
i: 8
```

Reply:
And even if no rehash, the order of insertion of each key might change the order of iterations.
```lua
local tl = {}
tl.a = 0
tl.b = 1
tl.c = 2
tl.d = 3
tl.e = 4
tl.f = 5
tl.g = 6
tl.h = 7
tl.i = 8

local tr = {}
tr.d = 3
tr.a = 0
tr.i = 8
tr.g = 6
tr.b = 1
tr.c = 2
tr.h = 7
tr.e = 4
tr.f = 5

for k, v in pairs(tl) do
    print(string.format("%s: %s", k, v))
end

print("==========================")

for k, v in pairs(tr) do
    print(string.format("%s: %s", k, v))
end
```

Reply:
Sure, but we will clear the cache when rebuild routes, so for one router the order of iteration is decided.

Reply:
changed to `ipairs`, but the resutl of `inst:get_fields()` which is implemented by Rust is still a hash table.

Reply:
> changed to `ipairs`, but the result of `inst:get_fields()` which is implemented by Rust is still a hash table.

I think that's ok, the `fields` is a module-level variable, It will not be changed throughout the life cycle of a specific worker process.